### PR TITLE
gadget: add gadget.ResolveContentPaths()

### PR DESCRIPTION
**Pull request by mvo5**
This function takes a gadget/kernel root dir and will resolve
all relative path from gadget into absolute paths and also
resolve any $kernel:tag/file-or-dir style references and
store them in Content.resolvedSource.

There will be a followup calls this before gadget updates
and removes the "contentDir" from the mountedfilesystem writer.

Based on #9776

### CODE REVIEW

File: gadget/layout.go [hunk 1]
```diff
@@ -240,6 +243,66 @@ func LayoutVolume(gadgetRootDir string, volume *Volume, constraints LayoutConstr
...
 
+// ResolveContentPaths resolves any "$kernel:" refs in the gadget
+// content and populates VolumeContent.resolvedSource with absolute
+// paths.
+//
+// XXX: maybe move into LayoutVolume(), operator on *Volume and make private?
+func ResolveContentPaths(gadgetRootDir, kernelRootDir string, lv *LaidOutVolume) error {
+	// Note that the kernelRootDir may reference the running
+	// kernel if there is a gadget update or the new kernel if
+	// there is a kernel update.
```
Comment (anonymouse64) [inline gadget/layout.go:254]: thanks for this comment

File: gadget/layout_test.go [hunk 2]
```diff
@@ -1136,3 +1138,119 @@ volumes:
...
+        filesystem-label: system-boot
+        size: 128M
+        content:
+          - source: $kernel:dtbs/boot-assets/
+            target: /
+          - source: $kernel:dtbs/some-file
+            target: /
+`
+
+func (p *layoutTestSuite) TestResolveContentPathsNotInWantedeAssets(c *C) {
```
Comment (anonymouse64) [inline gadget/layout_test.go:1180]:
```suggestion
func (p *layoutTestSuite) TestResolveContentPathsNotInWantedAssets(c *C) {
```

**Review by anonymouse64 — APPROVED** [review]

File: gadget/layout.go [hunk 3]
```diff
@@ -240,6 +243,67 @@ func LayoutVolume(gadgetRootDir string, volume *Volume, constraints LayoutConstr
...
+	for i := range lv.Volume.Structure {
+		if err := resolveContentPathsForStructure(gadgetRootDir, kernelRootDir, kernelInfo, &lv.Volume.Structure[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func resolveContentPathsForStructure(gadgetRootDir, kernelRootDir string, kernelInfo *kernel.Info, ps *VolumeStructure) error {
```
Comment (pedronis) [inline gadget/layout.go:268]: VolumeStructure I think should be the first argument

File: gadget/layout.go [hunk 4]
```diff
@@ -240,6 +243,67 @@ func LayoutVolume(gadgetRootDir string, volume *Volume, constraints LayoutConstr
...
+				newSource += "/"
+			}
+			ps.Content[i].resolvedSource = newSource
+		}
+	}
+
+	return nil
+}
+
+func resolveOne(gadgetRootDir, kernelRootDir string, kernelInfo *kernel.Info, pathOrRef string) (string, error) {
```
Comment (pedronis) [inline gadget/layout.go:287]: this maybe should be called resolveContentOne ?

File: gadget/layout.go [hunk 5]
```diff
@@ -240,6 +243,67 @@ func LayoutVolume(gadgetRootDir string, volume *Volume, constraints LayoutConstr
...
+		if !ok {
+			return "", fmt.Errorf("cannot find %q in kernel info from %q", wantedAsset, kernelRootDir)
+		}
+		if !strutil.ListContains(kernelAsset.Content, wantedContent) {
+			return "", fmt.Errorf("cannot find wanted kernel content %q in %q", wantedContent, kernelRootDir)
+		}
+		return filepath.Join(kernelRootDir, wantedContent), nil
+	}
+
+	return filepath.Join(gadgetRootDir, pathOrRef), nil
```
Comment (pedronis) [inline gadget/layout.go:304]: the fallback path doesn't seem to be tested

**Review by pedronis — COMMENTED** [review]
did a pass

File: gadget/layout.go (hunk 5 above)
Comment (mvo5) [reply gadget/layout.go:304]: added

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**